File: frontend/src/api/admin/tlsFingerprintProfile.ts
/**
 * Admin TLS Fingerprint Profile API endpoints
 * Handles TLS fingerprint profile CRUD for administrators
 */

import { apiClient } from '../client'

/**
 * TLS fingerprint profile interface
 */
export interface TLSFingerprintProfile {
  id: number
  platform: string
  name: string
  description: string | null
  user_agent: string
  enable_grease: boolean
  cipher_suites: number[]
  curves: number[]
  point_formats: number[]
  signature_algorithms: number[]
  alpn_protocols: string[]
  supported_versions: number[]
  key_share_groups: number[]
  psk_modes: number[]
  extensions: number[]
  compress_cert_algos: number[]
  delegated_credentials_algorithms: number[]
  application_settings_protocols: string[]
  created_at: string
  updated_at: string
}

/**
 * Create profile request
 */
export interface CreateProfileRequest {
  platform?: string
  name: string
  description?: string | null
  user_agent?: string
  enable_grease?: boolean
  cipher_suites?: number[]
  curves?: number[]
  point_formats?: number[]
  signature_algorithms?: number[]
  alpn_protocols?: string[]
  supported_versions?: number[]
  key_share_groups?: number[]
  psk_modes?: number[]
  extensions?: number[]
  compress_cert_algos?: number[]
  delegated_credentials_algorithms?: number[]
  application_settings_protocols?: string[]
}

/**
 * Update profile request
 */
export interface UpdateProfileRequest {
  platform?: string
  name?: string
  description?: string | null
  user_agent?: string
  enable_grease?: boolean
  cipher_suites?: number[]
  curves?: number[]
  point_formats?: number[]
  signature_algorithms?: number[]
  alpn_protocols?: string[]
  supported_versions?: number[]
  key_share_groups?: number[]
  psk_modes?: number[]
  extensions?: number[]
  compress_cert_algos?: number[]
  delegated_credentials_algorithms?: number[]
  application_settings_protocols?: string[]
}

export interface TLSFingerprintCaptureTask {
  id: number
  name: string
  status: 'running' | 'completed' | 'stopped' | string
  token?: string
  targets: Record<string, number>
  counts: Record<string, number>
  ua_keywords: string[]
  created_at: string
  updated_at: string
  completed_at?: string | null
}

export interface TLSFingerprintCaptureSample {
  id: number
  task_id: number
  platform: string
  user_agent: string
  fingerprint_hash: string
  profile: TLSFingerprintProfile
  raw_payload: string
  created_at: string
}

export interface StartCaptureTaskRequest {
  name?: string
  targets: Record<string, number>
  ua_keywords?: string[]
}

export interface SubmitCaptureRequest {
  token: string
  platform: string
  user_agent: string
  payload: string
}

export interface CaptureSubmitResult {
  accepted: boolean
  duplicate: boolean
  ignored_reason?: string
  fingerprint_hash?: string
  task: TLSFingerprintCaptureTask
  sample?: TLSFingerprintCaptureSample
  counts: Record<string, number>
}

export interface ImportCaptureTaskSamplesRequest {
  sample_ids?: number[]
}

export interface ImportCaptureRecord {
  profile: TLSFingerprintProfile
  duplicate: boolean
  fingerprint_hash: string
}

export interface ImportCaptureResult {
  imported: number
  duplicates: number
  profiles: ImportCaptureRecord[]
}

export async function list(): Promise<TLSFingerprintProfile[]> {
  const { data } = await apiClient.get<TLSFingerprintProfile[]>('/admin/tls-fingerprint-profiles')
  return data
}

export async function getById(id: number): Promise<TLSFingerprintProfile> {
  const { data } = await apiClient.get<TLSFingerprintProfile>(`/admin/tls-fingerprint-profiles/${id}`)
  return data
}

export async function create(profileData: CreateProfileRequest): Promise<TLSFingerprintProfile> {
  const { data } = await apiClient.post<TLSFingerprintProfile>('/admin/tls-fingerprint-profiles', profileData)
  return data
}

export async function update(id: number, updates: UpdateProfileRequest): Promise<TLSFingerprintProfile> {
  const { data } = await apiClient.put<TLSFingerprintProfile>(`/admin/tls-fingerprint-profiles/${id}`, updates)
  return data
}

export async function deleteProfile(id: number): Promise<{ message: string }> {
  const { data } = await apiClient.delete<{ message: string }>(`/admin/tls-fingerprint-profiles/${id}`)
  return data
}

export async function listCaptureTasks(): Promise<TLSFingerprintCaptureTask[]> {
  const { data } = await apiClient.get<TLSFingerprintCaptureTask[]>('/admin/tls-fingerprint-profiles/capture-tasks')
  return data
}

export async function startCaptureTask(request: StartCaptureTaskRequest): Promise<TLSFingerprintCaptureTask> {
  const { data } = await apiClient.post<TLSFingerprintCaptureTask>('/admin/tls-fingerprint-profiles/capture-tasks', request)
  return data
}

export async function getCaptureTask(id: number): Promise<TLSFingerprintCaptureTask> {
  const { data } = await apiClient.get<TLSFingerprintCaptureTask>(`/admin/tls-fingerprint-profiles/capture-tasks/${id}`)
  return data
}

export async function stopCaptureTask(id: number): Promise<TLSFingerprintCaptureTask> {
  const { data } = await apiClient.post<TLSFingerprintCaptureTask>(`/admin/tls-fingerprint-profiles/capture-tasks/${id}/stop`)
  return data
}

export async function listCaptureSamples(taskId: number): Promise<TLSFingerprintCaptureSample[]> {
  const { data } = await apiClient.get<TLSFingerprintCaptureSample[]>(`/admin/tls-fingerprint-profiles/capture-tasks/${taskId}/samples`)
  return data
}

export async function importCaptureTaskSamples(
  taskId: number,
  request: ImportCaptureTaskSamplesRequest = {}
): Promise<ImportCaptureResult> {
  const { data } = await apiClient.post<ImportCaptureResult>(
    `/admin/tls-fingerprint-profiles/capture-tasks/${taskId}/import`,
    request
  )
  return data
}

export async function submitCapture(request: SubmitCaptureRequest): Promise<CaptureSubmitResult> {
  const { data } = await apiClient.post<CaptureSubmitResult>('/tls-fingerprint-captures/submit', request)
  return data
}

export async function submitCaptureToEndpoint(
  endpoint: string,
  request: SubmitCaptureRequest
): Promise<CaptureSubmitResult> {
  const { data } = await apiClient.post<CaptureSubmitResult>(endpoint, request)
  return data
}

export const tlsFingerprintProfileAPI = {
  list,
  getById,
  create,
  update,
  delete: deleteProfile,
  listCaptureTasks,
  startCaptureTask,
  getCaptureTask,
  stopCaptureTask,
  listCaptureSamples,
  importCaptureTaskSamples,
  submitCapture,
  submitCaptureToEndpoint
}

export default tlsFingerprintProfileAPI

File: backend/internal/model/tls_fingerprint_profile.go
// Package model 定义服务层使用的数据模型。
package model

import (
	"time"

	"github.com/Wei-Shaw/sub2api/internal/pkg/tlsfingerprint"
)

// TLSFingerprintProfile TLS 指纹配置模板
// 包含完整的 ClientHello 参数，用于模拟特定客户端的 TLS 握手特征
type TLSFingerprintProfile struct {
	ID                             int64     `json:"id"`
	Platform                       string    `json:"platform"`
	Name                           string    `json:"name"`
	Description                    *string   `json:"description"`
	UserAgent                      string    `json:"user_agent"`
	EnableGREASE                   bool      `json:"enable_grease"`
	CipherSuites                   []uint16  `json:"cipher_suites"`
	Curves                         []uint16  `json:"curves"`
	PointFormats                   []uint16  `json:"point_formats"`
	SignatureAlgorithms            []uint16  `json:"signature_algorithms"`
	ALPNProtocols                  []string  `json:"alpn_protocols"`
	SupportedVersions              []uint16  `json:"supported_versions"`
	KeyShareGroups                 []uint16  `json:"key_share_groups"`
	PSKModes                       []uint16  `json:"psk_modes"`
	Extensions                     []uint16  `json:"extensions"`
	CompressCertAlgos              []uint16  `json:"compress_cert_algos"`
	DelegatedCredentialsAlgorithms []uint16  `json:"delegated_credentials_algorithms"`
	ApplicationSettingsProtocols   []string  `json:"application_settings_protocols"`
	CreatedAt                      time.Time `json:"created_at"`
	UpdatedAt                      time.Time `json:"updated_at"`
}

// Validate 验证模板配置的有效性
func (p *TLSFingerprintProfile) Validate() error {
	if p.Name == "" {
		return &ValidationError{Field: "name", Message: "name is required"}
	}
	if len(p.Platform) > 50 {
		return &ValidationError{Field: "platform", Message: "platform is too long"}
	}
	return nil
}

// ToTLSProfile 将领域模型转换为运行时使用的 tlsfingerprint.Profile
// 空切片字段会在 dialer 中 fallback 到内置默认值
func (p *TLSFingerprintProfile) ToTLSProfile() *tlsfingerprint.Profile {
	return &tlsfingerprint.Profile{
		Name:                           p.Name,
		UserAgent:                      p.UserAgent,
		EnableGREASE:                   p.EnableGREASE,
		CipherSuites:                   p.CipherSuites,
		Curves:                         p.Curves,
		PointFormats:                   p.PointFormats,
		SignatureAlgorithms:            p.SignatureAlgorithms,
		ALPNProtocols:                  p.ALPNProtocols,
		SupportedVersions:              p.SupportedVersions,
		KeyShareGroups:                 p.KeyShareGroups,
		PSKModes:                       p.PSKModes,
		Extensions:                     p.Extensions,
		CompressCertAlgos:              p.CompressCertAlgos,
		DelegatedCredentialsAlgorithms: p.DelegatedCredentialsAlgorithms,
		ApplicationSettingsProtocols:   p.ApplicationSettingsProtocols,
	}
}
